File: advent2017/6-2.pl
#!/usr/bin/perl

use List::Util qw(max);

@blocks = @ARGV;
@history = (join(' ', @blocks));
$step = 0;

print "$step: $history[0]\n";
sub step {
  $max = max @blocks;
  $bank = (grep { $blocks[$_] == $max } 0..$#blocks)[0];

  $blocks = $blocks[$bank];
  print "bank=$bank blocks=$blocks n=$#blocks\n";
  $blocks[$bank] = 0;
  for ($i = 1; $i <= $blocks; $i++) {
    $k = ($i + $bank) % @blocks;
    print "n=$#blocks i=$i k=$k\n";
    $blocks[$k]++;
    print join(' ', @blocks) . "\n";
  }

  $hash = join(' ', @blocks);
  $step++;

  print "$step: $hash\n";

  if ($found = (grep { $hash eq $_ } @history)[0]) {
    return $found;
  };

  push @history, $hash;

  return step();
}

$found_hash = step();
print "repeat: $found_hash";
print "steps: $step\n";
$index = (grep { $history[$_] eq $found_hash } 0..$#history)[0];
print "found at: $index\n";
print "diff " . ($step - $index) . "\n";

File: advent2017/6-2.c
#include <stdio.h>
#include <stdlib.h>

unsigned short *blocks;
unsigned short nblocks;
unsigned short **history;
unsigned short steps = 0;

unsigned int step() {
  unsigned short idx = 1, j;
  unsigned int i;
  unsigned short max, bank;
  unsigned short *hash = history[steps];

  printf("ref: %d\n", (int) hash);
  for(i = 0; i < nblocks; i++) {
    hash[i] = blocks[i];
    if(blocks[i] > max) {
      max = blocks[i];
      bank = i;
    }
  }

  for(i = 0; i < steps; i++) {
    for(j = 0; j < nblocks; j++) {
      if(history[i * nblocks][j] != hash[j]) {
        break;
      }
    }
    if(j == nblocks) {
      return i;
    }
  }

  blocks[bank] = 0;

  for(idx = 1; idx <= max; idx++) {
    blocks[(idx + bank) % nblocks]++;
  }

  steps++;
  return step();
}

int main(int argc, char **argv) {
  nblocks = argc;
  blocks = malloc(nblocks * sizeof(short));
  history = malloc(nblocks * sizeof(short) * 20000 * 10);

  if(!history) {
    printf("no memory :(\n");
    return 1;
  }

  unsigned short i = 0;

  for(i = 0; i < argc; i++) {
    blocks[i] = atoi(argv[1+i]);
  }

  unsigned int found = step();

  printf("steps: %d   found: %d   diff: %d", steps, found, steps - found);

  return 0;
}
